doc/administrator/configure.rst
.. SPDX-FileCopyrightText: 2017-present Tobias Kunze
.. SPDX-License-Identifier: CC-BY-SA-4.0

.. _configure:

Configuration
=============

You can configure imanage using config files or environment variables. You can
combine those two options, and their precedence is in this order:

1. Environment variables
2. Configuration files

   * The config file in the environment variable ``IMANAGE_CONFIG_FILE`` if present, **or**

   * The following three configuration files in this order:

      - The config file ``imanage.cfg`` in the ``src`` directory, next to the ``imanage.example.cfg`` file.
      - The config file ``~/.imanage.cfg`` in the home of the executing user.
      - The config file ``/etc/imanage/imanage.cfg``

3. Sensible defaults

This page explains the options grouped by configuration file sections. You’ll
find the environment variable next to their corresponding settings.  A config
file looks like this:

.. literalinclude:: ../../src/imanage.example.cfg
   :language: ini

The configuration format is the INI file format as understood by `Python`_.
It’s worth mentioning that you can use any of ``yes``/``no``, ``on``/``off``,
``true``/``false`` and ``1``/``0`` to set boolean settings. You can add add
comments by starting a line with ``;`` or ``#``. Also strings don’t need to be
quoted as every value is already a string.

The filesystem section
----------------------

``data``
~~~~~~~~

- The ``data`` option describes the path that is the base for all other
  directories. imanage will also save its log files there. Unless you have a
  compelling reason to keep other files apart, setting the ``data`` option is
  the easiest way to configure file storage.
- **Environment variable:** ``IMANAGE_DATA_DIR``
- **Default:** A directory called ``data`` next to imanage’s ``manage.py``.

``media``
~~~~~~~~~

- The ``media`` option sets the media directory that contains user generated files. It needs to
  be writable by the imanage process.
- **Environment variable:** ``IMANAGE_FILESYSTEM_MEDIA``
- **Default:** A directory called ``media`` in the ``data`` directory (see above).

``logs``
~~~~~~~~

- The ``logs`` option sets the log directory that contains logged data. It needs to
  be writable by the imanage process.
- **Environment variable:** ``IMANAGE_FILESYSTEM_LOGS``
- **Default:** A directory called ``logs`` in the ``data`` directory (see above).

``static``
~~~~~~~~~~

- The ``statics`` option sets the directory that contains static files. It needs to
  be writable by the imanage process. imanage will put files there during the ``rebuild`` and
  ``collectstatic`` commands.
- **Environment variable:** ``IMANAGE_FILESYSTEM_STATIC``
- **Default:** A directory called ``static.dist`` next to imanage’s ``manage.py``.

The site section
----------------

``debug``
~~~~~~~~~

- Decides if imanage runs in debug mode. Please use this mode for development and debugging, not
  for live usage. imanage behaves differently when ``debug`` is on, so you should run **all**
  your production commands, like e.g. ``rebuild`` with ``debug`` turned off.
- **Environment variable:** ``IMANAGE_DEBUG``
- **Default:** ``True`` if you’re executing ``runserver``, ``False`` otherwise. **Never run a
  production server in debug mode.**

``url``
~~~~~~~

- imanage uses this value when it has to render full URLs, for example in
  emails or feeds. It is also used to determined the allowed incoming hosts.
- **Environment variable:** ``IMANAGE_SITE_URL``
- **Default:** ``http://localhost``

``secret``
~~~~~~~~~~

- Every Django application has a secret that Django uses for cryptographic signing.
  You do not need to set this variable – imanage will generate a secret key and save it in a local file if
  you do not set it.
- **Default:** None

``media``
~~~~~~~~~

- The URL under which the media files (all files uploaded by users or generated by imanage) should be.
  If you change this value, remember to change your web server configuration, too!
- **Environment variable:** ``IMANAGE_SITE_MEDIA``
- **Default**: ``/media/``

``static``
~~~~~~~~~~

- The URL under which the static files should be. This includes all JavaScript and CSS files that
  are packaged with imanage. If you change this value, remember to change your web server configuration,
  too!
- **Environment variable:** ``IMANAGE_SITE_STATIC``
- **Default**: ``/static/``

``csp``, ``csp_script``, ``csp_style``, ``csp_img``, ``csp_form``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Use any of these settings to update the matching CSP security headers. Your
  configuration values will be *added* to the default headers, they won’t replace
  them completely!
  If you need to include special values like ``self``, remember to use single tick
  quotation marks: ``csp=https://rixx.de,'self'``
- **Environment variables:** ``IMANAGE_SITE_CSP``, ``IMANAGE_SITE_CSP_SCRIPT`` etc.
- **Default**: ``''``

``max_pagination_limit``
~~~~~~~~~~~~~~~~~~~~~~~~~

- The maximum number of items that can be requested per page. Users can request
  smaller page sizes, but cannot exceed this limit. Set to ``-1`` to remove the
  limit entirely. This setting applies to both API pages and UI pages.
- **Environment variable:** ``IMANAGE_MAX_PAGINATION_LIMIT``
- **Default:** ``250``

``highlighted_plugins``
~~~~~~~~~~~~~~~~~~~~~~~

- A comma-separated list of plugin module names that should be highlighted as
  recommended plugins. Highlighted plugins appear first in plugin lists and
  display a "Recommended" badge in both the plugin settings page and the event
  creation wizard. The order of the list does not matter, as the group of
  highlighted plugins will be sorted alphabetically.
- **Environment variable:** ``IMANAGE_HIGHLIGHTED_PLUGINS``
- **Default:** ``''``

The database section
--------------------

``backend``
~~~~~~~~~~~

- imanage supports most SQL databases. You’ll need to install the appropriate
  Python library for each of them, as described in the table below.
- **Environment variable:** ``IMANAGE_DB_TYPE``
- **Default:** ``sqlite3``

+------------+----------------------+-----------------------+
| Database   | Configuration string | pip package           |
+============+======================+=======================+
| PostgresQL | ``postgresql``       | ``imanage[postgres]`` |
+------------+----------------------+-----------------------+
| SQLite     | ``sqlite3``          | ``imanage``           |
+------------+----------------------+-----------------------+

``name``
~~~~~~~~

- The database’s name.
- **Environment variable:** ``IMANAGE_DB_NAME``
- **Default:** ``''``

``user``
~~~~~~~~

- The database user, if applicable.
- **Environment variable:** ``IMANAGE_DB_USER``
- **Default:** ``''``

``password``
~~~~~~~~~~~~

- The database password, if applicable.
- **Environment variable:** ``IMANAGE_DB_PASS``
- **Default:** ``''``

``host``
~~~~~~~~

- The database host, or the socket location, as needed. For local PostgreSQL authentication, you can leave this variable empty.
- **Environment variable:** ``IMANAGE_DB_HOST``
- **Default:** ``''``

``port``
~~~~~~~~

- The database port, if applicable.
- **Environment variable:** ``IMANAGE_DB_PORT``
- **Default:** ``''``

The mail section
----------------

This section serves to configure the email server imanage will use to send
emails by default. Events can override these settings in the web interface.
However, even with event-specific settings, imanage will use these global
settings for non-event-specific emails like password resets (even when
triggered from an event context) and team invitations.

``from``
~~~~~~~~

- The fall-back sender address, e.g. for when imanage sends event independent emails.
- **Environment variable:** ``IMANAGE_MAIL_FROM``
- **Default:** ``admin@localhost``

``host``
~~~~~~~~

- The email server host address.
- **Environment variable:** ``IMANAGE_MAIL_HOST``
- **Default:** ``localhost``

``port``
~~~~~~~~

- The email server port.
- **Environment variable:** ``IMANAGE_MAIL_PORT``
- **Default:** ``25``

``user``
~~~~~~~~

- The user account for mail server authentication, if needed.
- **Environment variable:** ``IMANAGE_MAIL_USER``
- **Default:** ``''``

``password``
~~~~~~~~~~~~

- The password for mail server authentication, if needed.
- **Environment variable:** ``IMANAGE_MAIL_PASSWORD``
- **Default:** ``''``

``tls``
~~~~~~~

- Should imanage use TLS when sending mail? Please choose either TLS or SSL.
- **Environment variable:** ``IMANAGE_MAIL_TLS``
- **Default:** ``False``

``ssl``
~~~~~~~

- Should imanage use SSL when sending mail? Please choose either TLS or SSL.
- **Environment variable:** ``IMANAGE_MAIL_SSL``
- **Default:** ``False``

The celery section
------------------

Celery is not a requirement for imanage. Celery runs as a separate process, and
allows you to execute long-running tasks away from the usual request-response
cycle.

.. warning:: If this config section is present, imanage will assume that Celery
             workers exist and collect talks. If you include this section without
             providing Celery workers, no asynchronous tasks (like email sending)
             will be processed. If you do not use Celery, omit this section in
             your configuration.

``backend``
~~~~~~~~~~~

- The celery backend. If you use a standard redis-based setup,
  ``redis://127.0.0.1/1`` would be a sensible value.
- **Environment variable:** ``IMANAGE_CELERY_BACKEND``
- **Default:** ``''``

``broker``
~~~~~~~~~~~

- The celery broker. If you use a standard redis-based setup,
  ``redis://127.0.0.1/2`` would be a sensible value.
- **Environment variable:** ``IMANAGE_CELERY_BROKER``
- **Default:** ``''``

The redis section
-----------------

If you configure a redis server, imanage can use it for locking, caching and
session storage to speed up operations. You will need to install ``django_redis``.

``location``
~~~~~~~~~~~~

- The location of redis, if you want to use it as a cache.
  ``redis://[:password]@127.0.0.1:6379/1`` would be a sensible value, or
  ``unix://[:password]@/path/to/socket.sock?db=0`` if you prefer to use sockets.
- **Environment variable:** ``IMANAGE_REDIS``
- **Default:** ``''``

``session``
~~~~~~~~~~~

- If you want to use redis as your session storage, set this to ``True``.
- **Environment variable:** ``IMANAGE_REDIS_SESSIONS``
- **Default:** ``False``

The logging section
-------------------

``email``
~~~~~~~~~

- The email address (or addresses, comma separated) to send system logs to.
- **Environment variable:** ``IMANAGE_LOGGING_EMAIL``
- **Default:** ``''``

``email_level``
~~~~~~~~~~~~~~~

- The log level to start sending emails at. Any of ``[DEBUG, INFO, WARNING, ERROR, CRITICAL]``.
- **Environment variable:** ``IMANAGE_LOGGING_EMAIL_LEVEL``
- **Default:** ``ERROR``

The locale section
------------------

``language_code``
~~~~~~~~~~~~~~~~~

- The system’s default locale.
- **Environment variable:** ``IMANAGE_LANGUAGE_CODE``
- **Default:** ``en``

``time_zone``
~~~~~~~~~~~~~

- The system’s default time zone as a ``pytz`` name.
- **Environment variable:** ``IMANAGE_TIME_ZONE``
- **Default:** ``UTC``

The files section
-----------------

``upload_limit``
~~~~~~~~~~~~~~~~

- The maximum file size for uploads in MB.
- **Environment variable:** ``IMANAGE_FILE_UPLOAD_LIMIT``
- **Default:** ``10``


.. _Python: https://docs.python.org/3/library/configparser.html


doc/changelog.rst
.. SPDX-FileCopyrightText: 2018-present Tobias Kunze
.. SPDX-License-Identifier: CC-BY-SA-4.0

.. _changelog:

Release Notes
=============

- :feature:`api` The review list endpoint now supports additional filters for submission state, pending state, track, submission type, and content locale.
- :feature:`orga` You can assign an identifier to custom fields, making it easier to use them in APIs and across events consistently. For choice fields, identifiers will be randomly assigned to each answer option as well to guarantee stable access.
- :bug:`schedule` The mobile view of the imanage schedule did not set a background colour, so the imanage schedule widget was hard to read when embedded on pages with a dark background colour.
- :bug:`api` When no review phases were active, organisers would not see any reviews in the API.
- :feature:`schedule` Instead of jumping directly to the current time, the imanage schedule (and schedule widget) now offer a “Jump to now” prompt instead.
- :bug:`orga:mail` The form to send reminders to open draft proposals was read-only and could not be used.
- :bug:`orga` Organisers could not add new review score settings, but could still modify existing score settings.
- :bug:`schedule` On the “featured sessions” page, only organisers were able to see speaker biographies in the featured sessions.
- :bug:`orga:schedule` When using the session form to schedule sessions rather than the schedule editor, organisers could schedule sessions outside the event dates.
- :announcement:`admin` imanage support now requires PostgreSQL 16 or higher.
- :feature:`orga` Organisers can now select a secondary sort column in list views.
- :feature:`cfp` Organisers can ask submitters to add tags to their proposals. Only public tags will be offered here; organisers can still use additional private tags.
- :feature:`orga:schedule,1326` Organisers can now create blocker sessions in the schedule editor. Blocker sessions work similar to breaks, but will never be public, and allow you to reserve schedule slots without accidentally moving another session there.
- :feature:`cfp` When logging in or changing your password, you can now toggle the plain-text display of the text you’re entering.
- :feature:`orga` Organisers can now select plugins when they create a new event, and if they copy another event’s settings, they can see and override the copied settings immediately.
- :feature:`schedule` There is now some basic spam protection for the open feedback form.
- :feature:`administrator,1947` Administrators can now send a test email to the server administrator email address on the administrator page, in order to verify that the instance-global email setup is working.
- :feature:`cfp` Users are now rate-limited on repeated failed login attempts.
- :feature:`cfp` If organisers set a colour with bad contrast to white font as their primary colour despite all warnings, imanage will render text on buttons and on the conference header in dark colours rather than in white.
- :bug:`orga:submissions,2220` Sessions without a track always showed with the anonymisation indicator in the organiser session list.
- :feature:`api` The submission organiser API now includes the ``created`` and ``updated`` timestamp fields. Please note that the ``updated`` timestamp is not guaranteed to be always set, particularly when only custom fields are being changes. If you need a reliable last modification date, please refer to the object changelog instead.
- :feature:`cfp` Organisers can now change the order of fields on each CfP page.
- :feature:`cfp` Organisers can now configure a maximum number of speakers in a proposal.
- :feature:`cfp` When speakers are invited as co-speaker to a proposal, the invite is now tracked and shown to the organisers.
- :announcement:`admin` imanage now requires Python 3.12 or newer.
- :feature:`api` At the root API URL, you will now find a link to the event API, as well as the latest API version that is available.
- :feature:`orga` When creating a new API token, you can now select and deselect whole rows and columns.
- :feature:`orga` You can now see which endpoints an existing API token can access.
- :bug:`orga:review` New review scores could only be created with a limited length at first, and could only be longer after having been saved once.
- :feature:`api` There is now an API for session feedback.
- :release:`2025.2.2 <2025-12-05>`
- :bug:`orga` There was a bug that prevented the saving of custom fields in the organiser area when they were set to be public.
- :release:`2025.2.1 <2025-12-03>`
- :bug:`orga:email` Emails containing schedule notifications for speakers in multilingual events could sometimes appear in mixed languages, with the general email text using the user’s preferred language and the notification section using the main event language.
- :bug:`api` Submission creation in the API was broken.
- :bug:`orga` There was a bug that prevented the saving of custom fields in the organiser area for some specific event settings.
- :bug:`orga:schedule,2209` Dragging sessions to their slots in the schedule editor did not work on MacOS browsers.
- :release:`2025.2.0 <2025-11-30>`
- :feature:`orga` Organisers can now limit who can see the answers to their custom field to specific organiser teams, e.g. for when sensitive information needs to be collected, but should not be visible to all organisers.
- :bug:`api` In the session slot API, the `end` field was required even though it was meant to be computed automatically by the session’s session type.
- :feature:`api` The session list in the API can now be filtered by track.
- :feature:`lang` New translation: Bulgarian!
- :bug:`orga` 'Filter by tracks' dropdown showed an incorrect submission count.
- :feature:`cfp` When users want to submit draft proposals, they will now be guided back through the CfP flow to make sure that all information is complete before final submission.

src/imanage/common/settings/config.py
# SPDX-FileCopyrightText: 2017-present Tobias Kunze
# SPDX-License-Identifier: AGPL-3.0-only WITH LicenseRef-Imanage-AGPL-3.0-Terms

import configparser
import os
import sys
from pathlib import Path

CONFIG = {
    "filesystem": {
        "base": {
            "default": Path(__file__).parent.parent.parent.parent,
        },
        "logs": {
            "default": None,
            "env": os.getenv("IMANAGE_FILESYSTEM_LOGS"),
        },
        "media": {
            "default": None,
            "env": os.getenv("IMANAGE_FILESYSTEM_MEDIA"),
        },
        "static": {
            "default": None,
            "env": os.getenv("IMANAGE_FILESYSTEM_STATIC"),
        },
    },
    "site": {
        "debug": {
            "default": "runserver" in sys.argv or "devserver" in sys.argv,
            "env": os.getenv("IMANAGE_DEBUG"),
        },
        "url": {
            "default": "http://localhost",
            "env": os.getenv("IMANAGE_SITE_URL"),
        },
        "https": {
            "env": os.getenv("IMANAGE_HTTPS"),
        },
        "cookie_domain": {
            "default": "",
            "env": os.getenv("IMANAGE_COOKIE_DOMAIN"),
        },
        "core_modules": {
            "default": "",
            "env": os.getenv("IMANAGE_CORE_MODULES"),
        },
        "highlighted_plugins": {
            "default": "",
            "env": os.getenv("IMANAGE_HIGHLIGHTED_PLUGINS"),
        },
        "media": {"default": "/media/", "env": os.getenv("IMANAGE_SITE_MEDIA")},
        "static": {"default": "/static/", "env": os.getenv("IMANAGE_SITE_STATIC")},
        "csp": {"default": "", "env": os.getenv("IMANAGE_SITE_CSP")},
        "csp_script": {"default": "", "env": os.getenv("IMANAGE_SITE_CSP_SCRIPT")},
        "csp_style": {"default": "", "env": os.getenv("IMANAGE_SITE_CSP_STYLE")},
        "csp_img": {"default": "", "env": os.getenv("IMANAGE_SITE_CSP_IMG")},
        "csp_form": {"default": "", "env": os.getenv("IMANAGE_SITE_CSP_FORM")},
        "max_pagination_limit": {
            "default": 250,
            "env": os.getenv("IMANAGE_MAX_PAGINATION_LIMIT"),
        },
    },
    "database": {
        "backend": {
            "default": "sqlite3",
            "env": os.getenv("IMANAGE_DB_TYPE"),
        },
        "name": {
            "env": os.getenv("IMANAGE_DB_NAME"),
        },
        "user": {
            "default": "",
            "env": os.getenv("IMANAGE_DB_USER"),
        },
        "password": {
            "default": "",
            "env": os.getenv("IMANAGE_DB_PASS"),
        },
        "host": {
            "default": "",
            "env": os.getenv("IMANAGE_DB_HOST"),
        },
        "port": {
            "default": "",
            "env": os.getenv("IMANAGE_DB_PORT"),
        },
    },
    "mail": {
        "from": {
            "default": "admin@localhost",
            "env": os.getenv("IMANAGE_MAIL_FROM"),
        },
        "host": {
            "default": "localhost",
            "env": os.getenv("IMANAGE_MAIL_HOST"),
        },
        "port": {
            "default": "25",
            "env": os.getenv("IMANAGE_MAIL_PORT"),
        },
        "user": {
            "default": "",
            "env": os.getenv("IMANAGE_MAIL_USER"),
        },
        "password": {
            "default": "",
            "env": os.getenv("IMANAGE_MAIL_PASSWORD"),
        },
        "tls": {
            "default": "False",
            "env": os.getenv("IMANAGE_MAIL_TLS"),
        },
        "ssl": {
            "default": "False",
            "env": os.getenv("IMANAGE_MAIL_SSL"),
        },
    },
    "redis": {
        "location": {
            "default": "False",
            "env": os.getenv("IMANAGE_REDIS"),
        },
        "session": {
            "default": "True",
            "env": os.getenv("IMANAGE_REDIS_SESSIONS"),
        },
    },
    "celery": {
        "broker": {
            "default": "",
            "env": os.getenv("IMANAGE_CELERY_BROKER"),
        },
        "backend": {
            "default": "",
            "env": os.getenv("IMANAGE_CELERY_BACKEND"),
        },
    },
    "logging": {
        "email": {
            "default": "",
            "env": os.getenv("IMANAGE_LOGGING_EMAIL"),
        },
        "email_level": {
            "default": "",
            "env": os.getenv("IMANAGE_LOGGING_EMAIL_LEVEL"),
        },
    },
    "locale": {
        "language_code": {
            "default": "en",
            "env": os.getenv("IMANAGE_LANGUAGE_CODE"),
        },
        "time_zone": {
            "default": "UTC",
            "env": os.getenv("IMANAGE_TIME_ZONE"),
        },
    },
    "files": {
        "upload_limit": {
            "default": 10,
            "env": os.getenv("IMANAGE_FILE_UPLOAD_LIMIT"),
        }
    },
}


def read_config_files(config):
    if "IMANAGE_CONFIG_FILE" in os.environ:
        with open(os.environ.get("IMANAGE_CONFIG_FILE"), encoding="utf-8") as fp:
            config_files = config.read_file(fp)
    else:
        config_files = config.read(
            [
                "/etc/imanage/imanage.cfg",
                Path.home() / ".imanage.cfg",
                "imanage.cfg",
            ],
            encoding="utf-8",
        )
    # .read() returns None if there are no config files
    return (
        config,
        config_files or [],
    )


def reduce_dict(data):
    return {
        section_name: {
            key: value for key, value in section_content.items() if value is not None
        }
        for section_name, section_content in data.items()
    }


def read_layer(layer_name, config):
    config_dict = reduce_dict(
        {
            section_name: {
                key: value.get(layer_name) for key, value in section_content.items()
            }
            for section_name, section_content in CONFIG.items()
        }
    )
    config.read_dict(config_dict)
    return config


def build_config():
    config = configparser.RawConfigParser()
    config = read_layer("default", config)
    config, config_files = read_config_files(config)
    config = read_layer("env", config)
    return config, config_files
